allennlp/modules/token_embedders/elmo_token_embedder.py

```diff
@@ -0,0 +1,245 @@
+'''
+Compute the context insensitive token embeddings from pretrained biLMs.
+'''
+import json
+import logging
+
+from typing import Dict
+
+import numpy
+import h5py
+
+from overrides import overrides
+
+import torch
+from torch.autograd import Variable
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.file_utils import cached_path
+from allennlp.modules.token_embedders.token_embedder import TokenEmbedder
+from allennlp.modules.highway import Highway
+from allennlp.nn.util import add_bos_eos
+from allennlp.data.token_indexers.elmo_indexer import ELMoCharacterMapper
+from allennlp.data import Vocabulary
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+DTYPE = 'float32'
+
+# pylint: disable=attribute-defined-outside-init
+
+
+@TokenEmbedder.register("elmo_token_embedder")
+class ELMoTokenEmbedder(TokenEmbedder):
+    '''
+    Compute context sensitive token representation using pretrained biLM.
+
+    This embedder has input character ids of size (batch_size, sequence_length, 50)
+    and returns (batch_size, sequence_length + 2, embedding_dim), where embedding_dim
+    is specified in the options file (typically 512).
+
+    We add special entries at the beginning and end of each sequence corresponding
+    to <S> and </S>, the beginning and end of sentence tokens.
+
+    Parameters
+    ----------
+    options_file : int
+        ELMo JSON options file
+    weight_file : int
+        ELMo hdf5 weight file
+    '''
+    def __init__(self,
+                 options_file: str,
+                 weight_file: str) -> None:
+        super(ELMoTokenEmbedder, self).__init__()
+
+        with open(cached_path(options_file), 'r') as fin:
+            self._options = json.load(fin)
+
+        self._weight_file = weight_file
+
+        self.output_dim = self._options['lstm']['projection_dim']
+
+        self._load_weights()
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self.output_dim
+
+    @overrides
+    def forward(self, inputs: torch.Tensor) -> Dict[str, torch.Tensor]:  # pylint: disable=arguments-differ
```

Comment:
Do you think it would be useful to have a `remove_beginning_and_end_tokens` flag which was by default false? Do you think people will want to use the CNN only part of the elmo embeddings independently? I guess if that is the case, we could have a `Seq2VecEncoder` which did that, so maybe not.

Reply:
Maybe, but we shouldn't add it until needed.